# Retail Sales, Returns and Shipping Analysis

## Business Context
A retail company wants to better understand its sales performance, product returns,
and shipping patterns in order to identify operational issues and improvement opportunities.

## Objective
- Understand overall sales and returns trends
- Identify products and regions with high return rates
- Explore potential relationships between shipping behavior and returns
- Provide actionable business recommendations

## Dataset
Source: Kaggle – Retail Sales, Returns and Shipping Dataset


In [3]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
orders = pd.read_csv(
    "/content/drive/MyDrive/retail_case/orders.csv",
    encoding="latin1"
)


In [7]:
returns = pd.read_csv(
    "/content/drive/MyDrive/retail_case/return.csv",
    encoding="latin1"
)

shipping = pd.read_csv(
    "/content/drive/MyDrive/retail_case/shipping_cost.csv",
    encoding="latin1"
)


In [8]:
orders.head()


Row ID        Order ID  Order Date  Ship Date  Delivery Duration  \
0       1  CA-2016-152156       42682      42685                  3   
1       2  CA-2016-152156       42682      42685                  3   
2       3  CA-2016-138688       42533      42537                  4   
3       4  US-2015-108966       42288      42295                  7   
4       5  US-2015-108966       42288      42295                  7   

        Ship Mode Customer ID    Customer Name    Segment        Country  ...  \
0    Second Class    CG-12520      Claire Gute   Consumer  United States  ...   
1    Second Class    CG-12520      Claire Gute   Consumer  United States  ...   
2    Second Class    DV-13045  Darrin Van Huff  Corporate  United States  ...   
3  Standard Class    SO-20335   Sean O'Donnell   Consumer  United States  ...   
4  Standard Class    SO-20335   Sean O'Donnell   Consumer  United States  ...   

        Product ID         Category  Sub-Category  \
0  FUR-BO-10001798        Furniture     Bookcases   
1  FUR-CH-10000454        Furniture        Chairs   
2  OFF-LA-10000240  Office Supplies        Labels   
3  FUR-TA-10000577        Furniture        Tables   
4  OFF-ST-10000760  Office Supplies       Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount Discount Value    Profit        COGS  
0     0.00       0.000000   41.9136 -220.046400  
1     0.00       0.000000  219.5820 -512.358000  
2     0.00       0.000000    6.8714   -7.748600  
3     0.45    -430.909875 -383.0310 -909.698625  
4     0.20      -4.473600    2.5164  -15.378000  

[5 rows x 24 columns]

In [9]:
orders.shape

(9994, 24)

In [10]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 24 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Row ID             9994 non-null   int64  
 1   Order ID           9994 non-null   object 
 2   Order Date         9994 non-null   int64  
 3   Ship Date          9994 non-null   int64  
 4   Delivery Duration  9994 non-null   int64  
 5   Ship Mode          9994 non-null   object 
 6   Customer ID        9994 non-null   object 
 7   Customer Name      9994 non-null   object 
 8   Segment            9994 non-null   object 
 9   Country            9994 non-null   object 
 10  City               9994 non-null   object 
 11  State              9994 non-null   object 
 12  Postal Code        9994 non-null   int64  
 13  Region             9994 non-null   object 
 14  Product ID         9994 non-null   object 
 15  Category           9994 non-null   object 
 16  Sub-Category       9994 

In [11]:
orders.head(10)

Row ID        Order ID  Order Date  Ship Date  Delivery Duration  \
0       1  CA-2016-152156       42682      42685                  3   
1       2  CA-2016-152156       42682      42685                  3   
2       3  CA-2016-138688       42533      42537                  4   
3       4  US-2015-108966       42288      42295                  7   
4       5  US-2015-108966       42288      42295                  7   
5       6  CA-2014-115812       41799      41804                  5   
6       7  CA-2014-115812       41799      41804                  5   
7       8  CA-2014-115812       41799      41804                  5   
8       9  CA-2014-115812       41799      41804                  5   
9      10  CA-2014-115812       41799      41804                  5   

        Ship Mode Customer ID    Customer Name    Segment        Country  ...  \
0    Second Class    CG-12520      Claire Gute   Consumer  United States  ...   
1    Second Class    CG-12520      Claire Gute   Consumer  United States  ...   
2    Second Class    DV-13045  Darrin Van Huff  Corporate  United States  ...   
3  Standard Class    SO-20335   Sean O'Donnell   Consumer  United States  ...   
4  Standard Class    SO-20335   Sean O'Donnell   Consumer  United States  ...   
5  Standard Class    BH-11710  Brosina Hoffman   Consumer  United States  ...   
6  Standard Class    BH-11710  Brosina Hoffman   Consumer  United States  ...   
7  Standard Class    BH-11710  Brosina Hoffman   Consumer  United States  ...   
8  Standard Class    BH-11710  Brosina Hoffman   Consumer  United States  ...   
9  Standard Class    BH-11710  Brosina Hoffman   Consumer  United States  ...   

        Product ID         Category  Sub-Category  \
0  FUR-BO-10001798        Furniture     Bookcases   
1  FUR-CH-10000454        Furniture        Chairs   
2  OFF-LA-10000240  Office Supplies        Labels   
3  FUR-TA-10000577        Furniture        Tables   
4  OFF-ST-10000760  Office Supplies       Storage   
5  FUR-FU-10001487        Furniture   Furnishings   
6  OFF-AR-10002833  Office Supplies           Art   
7  TEC-PH-10002275       Technology        Phones   
8  OFF-BI-10003910  Office Supplies       Binders   
9  OFF-AP-10002892  Office Supplies    Appliances   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   
5  Eldon Expressions Wood and Plastic Desk Access...   48.8600        7   
6                                         Newell 322    7.2800        4   
7                     Mitel 5320 IP Phone VoIP phone  907.1520        6   
8  DXL Angle-View Binders with Locking Rings by S...   18.5040        3   
9                   Belkin F5C206VTEL 6 Outlet Surge  114.9000        5   

  Discount Discount Value    Profit        COGS  
0     0.00       0.000000   41.9136 -220.046400  
1     0.00       0.000000  219.5820 -512.358000  
2     0.00       0.000000    6.8714   -7.748600  
3     0.45    -430.909875 -383.0310 -909.698625  
4     0.20      -4.473600    2.5164  -15.378000  
5     0.00       0.000000   14.1694  -34.690600  
6     0.00       0.000000    1.9656   -5.314400  
7     0.20    -181.430400   90.7152 -635.006400  
8     0.20      -3.700800    5.7825   -9.020700  
9     0.00       0.000000   34.4700  -80.430000  

[10 rows x 24 columns]

In [12]:
orders['Order ID'].nunique()


5009

In [13]:
len(orders)

9994

## Orders Table – Initial Observations

- The Orders table contains 9,994 rows and 5,009 unique Order IDs.
- This indicates that the dataset is at the order-line (product-level) granularity,
  where a single order can appear in multiple rows.
- Order-level metrics (e.g., number of orders) should therefore be calculated using unique Order IDs rather than row counts.



In [14]:
returns.shape

(296, 2)

## Returns Table – Initial Observations

- The Returns table contains 296 rows, which is significantly fewer than the total number of unique orders (5,009).
- This indicates that the table only includes orders that were returned.
- Orders not present in the Returns table are assumed to be non-returned orders.
- Therefore, the Returns table should be joined to the Orders table using a LEFT JOIN, and used to create a return indicator (flag).


In [15]:
returns.head()

Returned        Order ID
0      Yes  CA-2017-153822
1      Yes  CA-2017-129707
2      Yes  CA-2014-152345
3      Yes  CA-2015-156440
4      Yes  US-2017-155999

In [16]:
orders_with_returns = orders.merge(
    returns,
    on="Order ID",
    how="left"
)


In [20]:
orders_with_returns.head()


Row ID        Order ID  Order Date  Ship Date  Delivery Duration  \
0       1  CA-2016-152156       42682      42685                  3   
1       2  CA-2016-152156       42682      42685                  3   
2       3  CA-2016-138688       42533      42537                  4   
3       4  US-2015-108966       42288      42295                  7   
4       5  US-2015-108966       42288      42295                  7   

        Ship Mode Customer ID    Customer Name    Segment        Country  ...  \
0    Second Class    CG-12520      Claire Gute   Consumer  United States  ...   
1    Second Class    CG-12520      Claire Gute   Consumer  United States  ...   
2    Second Class    DV-13045  Darrin Van Huff  Corporate  United States  ...   
3  Standard Class    SO-20335   Sean O'Donnell   Consumer  United States  ...   
4  Standard Class    SO-20335   Sean O'Donnell   Consumer  United States  ...   

          Category Sub-Category  \
0        Furniture    Bookcases   
1        Furniture       Chairs   
2  Office Supplies       Labels   
3        Furniture       Tables   
4  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount Discount Value    Profit        COGS  Returned  
0     0.00       0.000000   41.9136 -220.046400       NaN  
1     0.00       0.000000  219.5820 -512.358000       NaN  
2     0.00       0.000000    6.8714   -7.748600       NaN  
3     0.45    -430.909875 -383.0310 -909.698625       NaN  
4     0.20      -4.473600    2.5164  -15.378000       NaN  

[5 rows x 25 columns]

In [21]:
orders_with_returns["is_returned"] = orders_with_returns["Returned"].notna()


In [22]:
orders_with_returns["is_returned"].value_counts()


is_returned
False    9194
True      800
Name: count, dtype: int64

## Returns Flag Creation

- orders and returns tables were merged using a LEFT JOIN on Order ID.
- A binary return indicator (`is_returned`) was created to identify returned orders.
- Since the orders table is at order-line level, the return flag is applied to all line items belonging to a returned order.


In [23]:
order_level_returns = (
    orders_with_returns
    .groupby("Order ID")["is_returned"]
    .max()
    .reset_index()
)


In [24]:
order_level_returns["is_returned"].value_counts()


is_returned
False    4713
True      296
Name: count, dtype: int64

## Order-Level vs. Line-Level Returns

- Initial return counts were calculated at the order-line level.
- Since a single order can contain multiple line items, return rates should be calculated at the order level.
- An order-level return flag was created by aggregating line-level return indicators, marking an order as returned if any of its line items were returned.


## Order-level return rate

In [25]:
total_orders = order_level_returns.shape[0]
returned_orders = order_level_returns["is_returned"].sum()

return_rate = returned_orders / total_orders

return_rate


np.float64(0.05909363146336594)

In [26]:
round(return_rate * 100, 2)


np.float64(5.91)

## Overall Return Rate

- The overall order-level return rate is 5.91%.
- This indicates that approximately 6 out of every 100 orders are returned.
- While not excessively high, this return rate warrants further investigation
  to identify categories or products with disproportionately high returns.


# Category-Level Return Rate

In [27]:
orders_with_order_flag = orders.merge(
    order_level_returns,
    on="Order ID",
    how="left"
)


In [28]:
orders_with_order_flag[["Order ID", "Category", "is_returned"]].head()


Order ID         Category  is_returned
0  CA-2016-152156        Furniture        False
1  CA-2016-152156        Furniture        False
2  CA-2016-138688  Office Supplies        False
3  US-2015-108966        Furniture        False
4  US-2015-108966  Office Supplies        False

In [29]:
category_return_rate = (
    orders_with_order_flag
    .groupby("Category")
    .agg(
        total_orders=("Order ID", "nunique"),
        returned_orders=("is_returned", "sum")
    )
    .reset_index()
)


In [30]:
category_return_rate["return_rate"] = (
    category_return_rate["returned_orders"]
    / category_return_rate["total_orders"]
)

category_return_rate["return_rate_pct"] = (
    category_return_rate["return_rate"] * 100
).round(2)


In [31]:
category_return_rate.sort_values(
    "return_rate_pct",
    ascending=False
)


Category  total_orders  returned_orders  return_rate  \
1  Office Supplies          3742              473     0.126403   
2       Technology          1544              156     0.101036   
0        Furniture          1764              171     0.096939   

   return_rate_pct  
1            12.64  
2            10.10  
0             9.69

## Category-Level Return Rate Methodology

- Return rates are calculated at the order level.
- Orders may contain items from multiple categories.
- A returned order contributes to the return rate of each category it includes.
- This analysis reflects the likelihood of an order being returned when a given category is present.


## Category-Level Insights

- Office Supplies has the highest order-level return rate at 12.64%.
- Technology and Furniture follow with return rates of 10.10% and 9.69%, respectively.
- All category-level return rates are higher than the overall return rate of 5.91%,
  reflecting that returned orders may contain multiple categories.
- The results indicate that orders including Office Supplies are more likely
  to be returned compared to other categories.


# Product-Level Analysis

- KPI: Return rate
- Granularity: Order-level
- Dimension: Product

In [32]:
product_return_rate = (
    orders_with_order_flag
    .groupby("Product Name")
    .agg(
        total_orders=("Order ID", "nunique"),
        returned_orders=("is_returned", "sum")
    )
    .reset_index()
)


In [33]:
product_return_rate["return_rate"] = (
    product_return_rate["returned_orders"]
    / product_return_rate["total_orders"]
)

product_return_rate["return_rate_pct"] = (
    product_return_rate["return_rate"] * 100
).round(2)


In [38]:
product_return_rate.sort_values(
    "total_orders",
    ascending=False
).head(10)


Product Name  total_orders  \
1493                             Staple envelope            48   
1499                                     Staples            46   
537                            Easy-staple paper            46   
259                      Avery Non-Stick Binders            20   
1500                     Staples in misc. colors            19   
942                   KI Adjustable-Height Table            18   
1496                              Staple remover            18   
1511                     Storex Dura Pro Binders            17   
1497                  Staple-based wall hangings            16   
1450  Situations Contoured Folding Chairs, 4/Set            15   

      returned_orders  return_rate  return_rate_pct  
1493                4     0.083333             8.33  
1499                2     0.043478             4.35  
537                 1     0.021739             2.17  
259                 0     0.000000             0.00  
1500                0     0.000000             0.00  
942                 4     0.222222            22.22  
1496                3     0.166667            16.67  
1511                0     0.000000             0.00  
1497                3     0.187500            18.75  
1450                2     0.133333            13.33

In [39]:
product_return_rate["total_orders"].describe()


count    1850.000000
mean        5.397838
std         3.132481
min         1.000000
25%         3.000000
50%         5.000000
75%         7.000000
max        48.000000
Name: total_orders, dtype: float64

## Product-Level Insights

- Most products in the dataset have low order frequency, with a median of 5 orders per product.
- High-volume products (e.g., Staples, Staple envelope) show relatively low return rates,
  indicating healthy demand and low return risk.
- Several mid-volume products (15–20 orders) exhibit disproportionately high return rates,
  such as adjustable-height tables and staple-related accessories.
- These products represent potential operational or expectation-related issues and
  should be prioritized for further investigation.


# Shipping Cost Context

- Shipping cost data is available at the state level and does not vary by order.
- As such, shipping cost was not used to calculate return rates directly.
- However, higher return rates observed in bulky products (e.g., furniture)
  may be partially influenced by higher shipping costs in certain states.
- This suggests a potential operational factor that could be explored further
  with more granular shipping data.


# Executive Summary

This analysis examines return behavior in a retail dataset consisting of 5,009 unique orders and nearly 10,000 order-line records. The primary objective was to calculate an accurate return rate and identify potential drivers of returns across categories and products, while maintaining methodological consistency.

Returns were analyzed at the order level, as a single order may contain multiple products and even one returned item results in a returned order. Using an order-level return flag, the overall return rate was calculated as 5.91%, meaning approximately 6 out of every 100 orders resulted in a return. This provides a reliable baseline KPI for further analysis.
Category-level analysis revealed meaningful differences in return behavior. Orders containing Office Supplies exhibited the highest return rate (12.64%), followed by Technology (10.10%) and Furniture (9.69%). All category-level return rates exceeded the overall average, reflecting the fact that returned orders may include multiple categories. These results indicate that return risk is not evenly distributed and is more pronounced in orders involving Office Supplies.

At the product level, the dataset showed a wide product assortment with generally low order frequency per product (median of 5 orders). While high-volume products such as staples and envelopes demonstrated low return rates and stable demand, several mid-volume products (15–20 orders) showed disproportionately high return rates. Notably, certain furniture items and office accessories exhibited return rates exceeding 15–20%, making them strong candidates for further operational or product-level investigation.
Shipping cost data was available only at the state level and lacked order-level identifiers. As a result, shipping variables were not used directly in return rate calculations. However, the elevated return rates observed in bulky and furniture-related products suggest that logistics-related factors (e.g., shipping complexity, delivery experience, or customer expectations) may play a role and warrant deeper analysis with more granular shipping data.

Overall, this analysis demonstrates that while the aggregate return rate is moderate, return behavior varies significantly by category and product. Targeted investigation of high-risk categories and mid-volume, high-return products presents a clear opportunity for reducing returns and improving operational efficiency.